File: packages/mirascript/src/vm/lib/global/sequence/with.ts
import { VM_ARRAY_MAX_LENGTH } from '../../../../helpers/constants.js';
import { toNumber, toString } from '../../../../helpers/convert/index.js';
import { isVmArray, isVmRecord } from '../../../../helpers/types.js';
import { isArray, isInteger, isNaN } from '../../../../helpers/utils.js';
import { Element } from '../../../helpers.js';
import type { VmArray, VmConst, VmValue } from '../../../types/index.js';
import { VmLib, expectArrayOrRecord, expectConst, throwError } from '../../helpers.js';

const arrIndex = (index: NonNullable<VmConst>): number => {
    const idx = Math.trunc(toNumber(index, Number.NaN));
    if (isNaN(idx) || idx < 0) {
        throwError('Array index must be a non-negative integer', index);
    }
    if (idx >= VM_ARRAY_MAX_LENGTH) {
        throwError(`Array index exceeds maximum limit of ${VM_ARRAY_MAX_LENGTH}`, index);
    }
    return idx;
};

const isArrIndex = (key: NonNullable<VmConst>): key is number => {
    if (typeof key != 'number') return false;
    return isInteger(key) && key >= 0 && key < VM_ARRAY_MAX_LENGTH;
};

const withInner = (obj: VmConst | undefined, key: VmArray, keyIndex: number, value: VmConst): VmConst => {
    if (keyIndex >= key.length) {
        return value;
    }
    const k = key[keyIndex]!;
    let result: Array<VmConst | undefined> | Record<string, VmConst | undefined>;
    if (isVmArray(obj)) {
        result = [...obj];
    } else if (isVmRecord(obj)) {
        result = { ...obj };
    } else if (isArrIndex(k)) {
        result = [];
    } else {
        result = {};
    }
    if (isArray(result)) {
        const index = arrIndex(k);
        while (index > result.length) {
            result.push(null);
        }
        result[index] = withInner(result[index], key, keyIndex + 1, value);
    } else {
        const prop = toString(k, undefined);
        result[prop] = withInner(result[prop], key, keyIndex + 1, value);
    }
    return result;
};

const normalizeEntries = (data: VmConst, entries: Array<VmValue | undefined>): Map<NonNullable<VmConst>, VmConst> => {
    if (entries.length % 2 !== 0) {
        throwError('Expected even number of entries', data);
    }
    const entryData = new Map<NonNullable<VmConst>, VmConst>();
    for (let i = 0; i < entries.length; i += 2) {
        let key = entries[i]!;
        expectConst('key', key, data);
        if (key == null) {
            continue;
        }
        if (isVmArray(key)) {
            if (key.length === 0 || key.includes(null) || key.includes(undefined)) {
                continue;
            } else if (key.length === 1) {
                key = key[0]!;
            }
        }
        const value = entries[i + 1]!;
        entryData.set(key, Element(value));
    }
    return entryData;
};

const _with = VmLib(
    (data, ...entries) => {
        expectArrayOrRecord('data', data, data);
        if (entries.length === 0) {
            return data;
        }

        const entryData = normalizeEntries(data, entries);
        if (isVmArray(data)) {
            const result: Array<VmConst | undefined> = [...data];
            for (const [key, value] of entryData) {
                let index: number;
                let val: VmConst;
                if (isVmArray(key)) {
                    index = arrIndex(key[0]!);
                    val = withInner(result[index], key, 1, value);
                } else {
                    index = arrIndex(key);
                    val = value;
                }
                while (index > result.length) {
                    result.push(null);
                }
                result[index] = val;
            }
            return result;
        } else {
            const result: Record<string, VmConst | undefined> = { ...data };
            for (const [key, value] of entryData) {
                let prop: string;
                let val: VmConst;
                if (isVmArray(key)) {
                    const firstKey = key[0]!;
                    prop = toString(firstKey, undefined);
                    val = withInner(result[prop], key, 1, value);
                } else {
                    prop = toString(key, undefined);
                    val = value;
                }
                result[prop] = val;
            }
            return result;
        }
    },
    {
        summary: '在数组或记录中设置多个键值对',
        params: {
            data: '要设置的数组或记录',
            '..entries': '要设置的键值对，成对出现',
        },
        paramsType: {
            data: 'array | record',
            '..entries': `[..[number | string | (number | string)[], any][]]`,
        },
        returnsType: 'type(data)',
        examples: [
            `with([10, 20], 2, 99, 3 ,100) // [10, 20, 99, 100]`,
            `(a: 1)::with(["b", 1], 2) // (a: 1, b: [nil, 2])`,
        ],
    },
);
export { _with as 'with' };

File: crates/python/mirascript/_vm/lib/vm_global/sequence/with_.py
from __future__ import annotations
import math
from typing_extensions import Sequence

from ....._helpers.checker import is_safe_integer, is_number
from ....._helpers.types import is_vm_array, is_vm_record
from ....._helpers.convert import to_number, to_string
from ....._helpers.constants import Uninitialized, VM_ARRAY_MAX_LENGTH
from ....operations.helpers import Element
from ....types import VmConst, VmAny
from ..._helpers import (
    _expect_array_or_record,
    _throw_error,
    _expect_const,
)


def _arr_index(index: VmConst) -> int:
    num = to_number(index, math.nan)
    if not math.isfinite(num) or num < 0:
        _throw_error("Array index must be a non-negative integer", index)
    idx = math.trunc(num)
    if idx >= VM_ARRAY_MAX_LENGTH:
        _throw_error(
            f"Array index exceeds maximum limit of {VM_ARRAY_MAX_LENGTH}", index
        )
    return idx


def _is_arr_index(key: VmConst) -> bool:
    if not is_number(key):
        return False
    if not is_safe_integer(key):
        return False
    return key >= 0 and key <= VM_ARRAY_MAX_LENGTH


def _with_inner(
    obj: VmConst, key: Sequence[VmConst], key_index: int, value: VmConst
) -> VmConst:
    if key_index >= len(key):
        return value

    k = key[key_index]
    result = None
    if is_vm_array(obj):
        result = obj.copy()
    elif is_vm_record(obj):
        result = obj.copy()
    elif _is_arr_index(k):
        result = []
    else:
        result = {}

    if is_vm_array(result):
        index = int(_arr_index(k))
        while index >= len(result):
            result.append(None)
        result[index] = _with_inner(result[index], key, key_index + 1, value)
    else:
        key_str = to_string(k)
        result[key_str] = _with_inner(
            result.get(key_str, None), key, key_index + 1, value
        )
    return result


def _normalize_entries(
    data: VmConst, entries: Sequence[VmAny]
) -> list[tuple[VmConst, VmConst]]:
    if len(entries) % 2 != 0:
        _throw_error(
            "with_ function requires even number of arguments as key-value pairs", data
        )

    entryData: list[tuple[VmConst, VmConst]] = []

    for i in range(0, len(entries), 2):
        key = _expect_const("key", entries[i], data)
        if key is None:
            continue

        if is_vm_array(key):
            if len(key) == 0 or None in key or Uninitialized in key:
                continue
            if len(key) == 1:
                key = key[0]

        value = entries[i + 1]
        entryData.append((key, Element(value)))

    return entryData


def with_(data: VmAny = Uninitialized, *args: VmAny) -> VmConst:
    data = _expect_array_or_record("data", data, [])
    if len(args) == 0:
        return data

    entryData = _normalize_entries(data, args)
    if is_vm_array(data):
        result = data.copy()
        for key, element in entryData:
            index: int = 0
            val = None

            if is_vm_array(key):
                index = _arr_index(key[0])
                if index < 0:
                    continue
                val = _with_inner(
                    result[index] if index < len(result) else None, key, 1, element
                )
            else:
                index = _arr_index(key)
                if index < 0:
                    continue
                val = element
            while index >= len(result):
                result.append(None)
            result[index] = val
        return result
    else:
        result = data.copy()
        for key, element in entryData:
            if is_vm_array(key):
                firstKey = key[0]
                prop = to_string(firstKey)
                val = _with_inner(result.get(prop, None), key, 1, element)
            else:
                prop = to_string(key)
                val = element
            result[prop] = val
        return result
